rust/wal3/src/reader.rs

```diff
@@ -0,0 +1,104 @@
+// NOTE(rescrv):  All caches align to storage.  For now, implement without caching.  Caching
+// should/could literally be a layer over storage, so add it later once correctness without caching
+// is ensured by adequate testing.
+
+use std::sync::Arc;
+
+use chroma_storage::Storage;
+
+use crate::{Error, Fragment, LogPosition, LogReaderOptions, Manifest, Snapshot};
+
+/// Limits allows encoding things like offset, timestamp, and byte size limits for the read.
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+pub struct Limits {
+    pub max_files: u64,
+    pub max_bytes: u64,
+}
+
+/// LogReader is a reader for the log.
+pub struct LogReader {
+    options: LogReaderOptions,
+    storage: Arc<Storage>,
+    prefix: String,
+}
+
+impl LogReader {
+    pub async fn open(
+        options: LogReaderOptions,
+        storage: Arc<Storage>,
+        prefix: String,
+    ) -> Result<Self, Error> {
+        Ok(Self {
+            options,
+            storage,
+            prefix,
+        })
+    }
+
+    pub async fn scan(&self, from: LogPosition, limits: Limits) -> Result<Vec<Fragment>, Error> {
+        let Some((manifest, _)) = Manifest::load(&self.storage, &self.prefix).await? else {
+            return Err(Error::UninitializedLog);
+        };
+        let mut snapshots = manifest
+            .snapshots
+            .iter()
+            .filter(|s| s.limit.offset() > from.offset())
```

Comment:
maybe document that this method is exclusive of the input offset?

Alternatively use something like https://doc.rust-lang.org/std/ops/struct.Range.html semantics